### PR TITLE
adding nvme related utilities

### PR DESCRIPTION
**Pull request by Naresh-ibm**
Added commonly used methods for nvme operations.
will be adding more to this utility in future which are needed for nvme operations.

Signed-off-by: Naresh Bannoth <nbannoth@in.ibm.com>

### CODE REVIEW

**Review by abdhaleegit — APPROVED** [review]
LGTM

File: avocado/utils/nvme.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    path = '/sys/bus/pci/devices/%s/nvme/' % pci_addr
+    return os.listdir(path)
+
+
+def get_number_of_ns_supported(controller_name):
+    """
+    Returns the number of namespaces supported for the nvme adapter
+
+    :param controller_name: Name of the controller eg: nvme0
+    :rtype: interger
```
Comment (abdhaleegit) [inline avocado/utils/nvme.py:51]: interger?

File: avocado/utils/nvme.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                      shell=True).stdout_text
+    return int(out.split(':')[-1].strip())
+
+
+def get_total_capacity(controller_name):
+    """
+    Returns the total capacity of the nvme adapter
+
+    :param controller_name: Name of the controller eg: nvme0
+    :rtype: interger
```
Comment (abdhaleegit) [inline avocado/utils/nvme.py:64]: Spell

File: avocado/utils/nvme.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    """
+    cmd = 'nvme id-ctrl /dev/%s | grep ^tnvmcap' % controller_name
```
Comment (abdhaleegit) [inline avocado/utils/nvme.py:66]: Possible to not use pipe?

File: avocado/utils/nvme.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+def get_controller_id(controll_name):
+    """
+    Returns the nvme controller id
+    """
+    cmd = "nvme list-ctrl %s" % controll_name
+    output = process.system_output(cmd, shell=True,
+                                   ignore_status=True).decode("utf-8")
+    for line in output.splitlines():
+        if '0]' in line:
```
Comment (abdhaleegit) [inline avocado/utils/nvme.py:80]: hope ] special character is formatted?

**Review by abdhaleegit — CHANGES_REQUESTED** [review]
LGTM

**Comment by abdhaleegit** [comment]
@Naresh-ibm also important modules should be sorted in order

File: avocado/utils/nvme.py (hunk 1 above)
Comment (Naresh-ibm) [reply avocado/utils/nvme.py:51]: sure , thanks. will change it

File: avocado/utils/nvme.py (hunk 3 above)
Comment (Naresh-ibm) [reply avocado/utils/nvme.py:66]: will explore more on this option

File: avocado/utils/nvme.py (hunk 4 above)
Comment (Naresh-ibm) [reply avocado/utils/nvme.py:80]: it is taken care

**Review by abdhaleegit — APPROVED** [review]
LGTM

File: avocado/utils/nvme.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+"""
+
+
+import os
+from avocado.utils import process
+
+
+class NvmeError(Exception):
+    """
+    Generic DiskError
```
Comment (PraveenPenguin) [inline avocado/utils/nvme.py:31]: please reword this as we are catching here Nvme error, just a suggestion can we narrow down Exception as it has broader scope

File: avocado/utils/nvme.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    """
+
```
Comment (PraveenPenguin) [inline avocado/utils/nvme.py:33]: remove this whit space

File: avocado/utils/nvme.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+def get_controller_name(pci_addr):
+    """
+    Returns the controller/Adapter name with the help of pci_address
+
+    :param pci_addr: pci_address of the adapter
+
+    """
+    path = '/sys/bus/pci/devices/%s/nvme/' % pci_addr
```
Comment (PraveenPenguin) [inline avocado/utils/nvme.py:42]: what if PCI device itself given wrong we shoule be handel those case gracefuly

File: avocado/utils/nvme.py (hunk 3 above)
Comment (PraveenPenguin) [reply avocado/utils/nvme.py:66]: yes this valid suggestion @Naresh-ibm please look on this

**Review by PraveenPenguin — CHANGES_REQUESTED** [review]

File: avocado/utils/nvme.py (hunk 6 above)
Comment (Naresh-ibm) [reply avocado/utils/nvme.py:33]: pylint wants two lines spaces here :)

**Review by abdhaleegit — APPROVED** [review]
LGTM

File: avocado/utils/nvme.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    :param controller_name: Name of the controller eg: nvme0
+    :rtype: integer
+    """
+    cmd = 'nvme id-ctrl /dev/%s' % controller_name
+    out = process.run(cmd, ignore_status=True, sudo=True,
+                      shell=True).stdout_text
+    for line in out.splitlines():
+        if line.split(":")[0].strip() == 'tnvmcap':
+            return int(line.split(':')[-1].strip())
```
Comment (abdhaleegit) [inline avocado/utils/nvme.py:77]: return empty if fail

File: avocado/utils/nvme.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    :param controller_name: Name of the controller eg: nvme0
+    :rtype: integer
+    """
+    cmd = 'nvme id-ctrl /dev/%s' % controller_name
+    out = process.run(cmd, ignore_status=True, sudo=True,
+                      shell=True).stdout_text
+    for line in out.splitlines():
+        if line.split(":")[0].strip() == 'nn':
+            return int(line.split(':')[-1].strip())
```
Comment (abdhaleegit) [inline avocado/utils/nvme.py:62]: return something if fail

**Review by abdhaleegit — CHANGES_REQUESTED** [review]

**Comment by abdhaleegit** [comment]
@PraveenPenguin we cab merge this ?

**Review by PraveenPenguin — APPROVED** [review]

**Comment by richtja** [comment]
Hi @Naresh-ibm and @PraveenPenguin, thank you for this work, unfortunately this merge broke static-checks on master. Can you please fix it? Thanks.
